# Raw RNA-Seq reads

The raw sequencing data generated for this project has been deposited in the Sequence Read Archive (SRA) and is available for download via the respective BioProject IDs: PRJNA939006 and PRJNA890248. Access to individual data files can be obtained through the SRA accessions SRR23625966 to SRR23626145 and SRR21891679 to SRR21891705.

# Analyses results files

To adhere to our commitment to reproducible and open science practices, we have made all the codes, scripts, and results utilized in this project available on GitLab. Access to these resources can be obtained through [here](https://gitlab.gwdg.de/armin.dadras/time-resolved-oxidative-signal-convergence-across-the-algae-embryophyte-divide).

# Annotation files that were used in this study

In the course of this study, we employed genome annotation and protein sequences of various species. Below is the comprehensive list of these resources and their respective locations for reference:

| Species | Paper | Downloaded from |
|:---------------|:------|:----------------|
| C. reinhardtii | Craig, Rory J., et al. "The Chlamydomonas Genome Project, version 6: Reference assemblies for mating-type plus and minus strains reveal extensive structural mutation in the laboratory." The Plant Cell 35.2 (2023): 644-672. | https://data.jgi.doe.gov/refine-download/phytozome?organism=CreinhardtiiCC-4532&expanded=707 |
| O. lucimarinus | Palenik, Brian, et al. "The tiny eukaryote Ostreococcus provides genomic insights into the paradox of plankton speciation." Proceedings of the National Academy of Sciences 104.18 (2007): 7705-7710. | https://data.jgi.doe.gov/refine-download/phytozome?q=Ostreococcus+lucimarinus&expanded=Phytozome-231 |
| M. viride | Liang, Zhe, et al. "Mesostigma viride genome and transcriptome provide insights into the origin and evolution of Streptophyta." Advanced Science 7.1 (2020): 1901850. | https://genome.jgi.doe.gov/portal/pages/dynamicOrganismDownload.jsf?organism=Mesvir1 |
| C. melkonianii | Wang, Sibo, et al. "Genomes of early-diverging streptophyte algae shed light on plant terrestrialization." Nature Plants 6.2 (2020): 95-106. | https://ftp.cngb.org/pub/CNSA/data1/CNP0000228/CNS0021447/CNA0002353/ |
| K. nitens | Hori, Koichi, et al. "Klebsormidium flaccidum genome reveals primary factors for plant terrestrial adaptation." Nature communications 5.1 (2014): 3978. | https://genome.jgi.doe.gov/portal/pages/dynamicOrganismDownload.jsf?organism=Klenit1 |
| C. braunii | Nishiyama, Tomoaki, et al. "The Chara genome: secondary complexity and implications for plant terrestrialization." Cell 174.2 (2018): 448-464. | https://bioinformatics.psb.ugent.be/gdb/Chara_braunii/ |
| A. agrestis oxford | Li, Fay-Wei, et al. "Anthoceros genomes illuminate the origin of land plants and the unique biology of hornworts." Nature plants 6.3 (2020): 259-272. | https://www.hornworts.uzh.ch/en/download.html |
| M. polymorpha | Montgomery, Sean A., et al. "Chromatin organization in early land plants reveals an ancestral association between H3K27me3, transposons, and constitutive heterochromatin." Current Biology 30.4 (2020): 573-588. | https://marchantia.info/download/MpTak_v6.1/ |
| P. patens | Lang, Daniel, et al. "The Physcomitrella patens chromosome‐scale assembly reveals moss genome structure and evolution." The Plant Journal 93.3 (2018): 515-533. | https://data.jgi.doe.gov/refine-download/phytozome?organism=Ppatens&expanded=318 |
| S. moellendorffii | Banks, Jo Ann, et al. "The Selaginella genome identifies genetic changes associated with the evolution of vascular plants." science 332.6032 (2011): 960-963. | https://data.jgi.doe.gov/refine-download/phytozome?q=Selaginella+moellendorffii&expanded=Phytozome-91 |
| A. filiculoides | Li, Fay-Wei, et al. "Fern genomes elucidate land plant evolution and cyanobacterial symbioses." Nature plants 4.7 (2018): 460-472. | https://fernbase.org/ftp/Azolla_filiculoides/Azolla_asm_v1.1/ |
| A. thaliana | Cheng, Chia‐Yi, et al. "Araport11: a complete reannotation of the Arabidopsis thaliana reference genome." The Plant Journal 89.4 (2017): 789-804. | https://data.jgi.doe.gov/refine-download/phytozome?q=Arabidopsis+thaliana&expanded=Phytozome-447 |
| S. lycopersicum | Hosmani, Prashant S., et al. "An improved de novo assembly and annotation of the tomato reference genome using single-molecule sequencing, Hi-C proximity ligation and optical maps." biorxiv (2019): 767764. | https://data.jgi.doe.gov/refine-download/phytozome?q=Solanum+lycopersicum&expanded=Phytozome-691 |
| Z. mays | Jiao, Yinping, et al. "Improved maize reference genome with single-molecule technologies." Nature 546.7659 (2017): 524-527. | https://data.jgi.doe.gov/refine-download/phytozome?q=zea+mays&expanded=Phytozome-493 |
| B. distachyon | DNA sequencing and assembly Barry Kerrie 5 Lucas Susan 5 Harmon-Smith Miranda 5 Lail Kathleen 5 Tice Hope 5 Schmutz (Leader) Jeremy 4 Grimwood Jane 4 McKenzie Neil 7 Bevan Michael W. michael. bevan@ bbsrc. ac. uk 7 k, Gene analysis and annotation Haberer Georg 16 Spannagl Manuel 16 Mayer (Leader) Klaus 16 Rattei Thomas 17 Mitros Therese 6 Rokhsar Dan 6 Lee Sang-Jik 18 Rose Jocelyn KC 18 Mueller Lukas A. 19 York Thomas L. 19, and Comparative genomics Salse (Leader) Jerome 27 Murat Florent 27 Abrouk Michael 27 Haberer Georg 16 Spannagl Manuel 16 Mayer Klaus 16 Bruggmann Remy 13 Messing Joachim 13 You Frank M. 8 Luo Ming-Cheng 8 Dvorak Jan 8. "Genome sequencing and analysis of the model grass Brachypodium distachyon." Nature 463.7282 (2010): 763-768. | https://data.jgi.doe.gov/refine-download/phytozome?q=Brachypodium+distachyon&expanded=Phytozome-556 |
| O. sativa | Ouyang, S. et al. The TIGR Rice Genome Annotation Resource: improvements and new features. Nucleic Acids Res. 35, D883–D887 (2007). | https://data.jgi.doe.gov/refine-download/phytozome?q=Oryza+sativa&expanded=Phytozome-323 |
| P. margaritaceum | Jiao, C. et al. The Penium margaritaceum genome: hallmarks of the origins of land plants. Cell 181, 1097–1111.e12 (2020). | http://bioinfo.bti.cornell.edu/cgi-bin/Penium/download.cgi |
| Closterium sp. NIES67 | Tsuchikane, Yuki, and Hiroyuki Sekimoto. "The genus Closterium, a new model organism to study sexual reproduction in streptophytes." New Phytologist 221.1 (2019): 99-104. | https://www.ncbi.nlm.nih.gov/protein/?term=BQMA+Closterium |
| Z. circumcarinatum SAG698-1b | Feng, Xuehuan, et al. "Chromosome-level genomes of multicellular algal sisters to land plants illuminate signaling network evolution." bioRxiv (2023): 2023-01. | https://phycocosm.jgi.doe.gov/Zygcir6981b_2/Zygcir6981b_2.home.html |
| Z. circumcarinatum SAG698-1a | Feng, Xuehuan, et al. "Chromosome-level genomes of multicellular algal sisters to land plants illuminate signaling network evolution." bioRxiv (2023): 2023-01. | https://phycocosm.jgi.doe.gov/Zygcyl6981a_1/Zygcyl6981a_1.home.html |
| Z. circumcarinatum UTEX1560 | Feng, Xuehuan, et al. "Chromosome-level genomes of multicellular algal sisters to land plants illuminate signaling network evolution." bioRxiv (2023): 2023-01. | https://phycocosm.jgi.doe.gov/Zygcir1560_1/Zygcir1560_1.home.html |
| Z. circumcarinatum UTEX1559 | Feng, Xuehuan, et al. "Chromosome-level genomes of multicellular algal sisters to land plants illuminate signaling network evolution." bioRxiv (2023): 2023-01. | https://phycocosm.jgi.doe.gov/Zygcir1559_1/Zygcir1559_1.home.html |
| M. endlicherianum | Dadras, Armin, et al. "Environmental gradients reveal stress hubs pre-dating plant terrestrialization." Nature Plants (2023): 1-20. | https://mesotaenium.uni-goettingen.de/download.html |
| S. muscicola | Cheng, Shifeng, et al. "Genomes of subaerial Zygnematophyceae provide insights into land plant evolution." Cell 179.5 (2019): 1057-1067. | https://figshare.com/articles/dataset/ |
| P. coloniale | Li, Linzhou, et al. "The genome of Prasinoderma coloniale unveils the existence of a third phylum within green plants." Nature ecology & evolution 4.9 (2020): 1220-1231. | https://phycocosm.jgi.doe.gov/Praco1/Praco1.home.html |
