# Methods and tools that were used in this research

| Tool | Version | Link |
|:---------------|:------|:----------------|
| ape | 5.7-1 | https://github.com/emmanuelparadis/ape |
| alluvial | 0.2-0 | https://github.com/mbojan/alluvial |
| Blast | 2.15.0+ | https://ncbiinsights.ncbi.nlm.nih.gov/2023/11/21/faster-focused-searches-blast-2-15/ |
| clusterProfiler | 4.10.0 | https://github.com/YuLab-SMU/clusterProfiler |
| ComplexHeatmap | 2.18.0 | https://github.com/jokergoo/ComplexHeatmap |
| conda | 4.12.0 | https://conda.io/projects/conda/en/latest/user-guide/install/index.html |
| cowplot | 1.1.2 | https://github.com/wilkelab/cowplot |
| Dirichlet process Gaussian process (DPGP) | - | https://github.com/PrincetonUniversity/DP_GP_cluster |
| DT | 0.31 | https://rstudio.github.io/DT/ |
| edgeR | 4.0.6 | https://bioconductor.org/packages/release/bioc/html/edgeR.html |
| eggNOG-mapper | 2.1.12 | https://github.com/eggnogdb/eggnog-mapper |
| enrichplot | 1.22.0 | https://github.com/YuLab-SMU/enrichplot |
| FastQC | 0.12.1 | https://www.bioinformatics.babraham.ac.uk/projects/fastqc/ |
| ggdist | 3.3.1 | https://mjskay.github.io/ggdist/ |
| ggforce | 0.4.1 | https://github.com/thomasp85/ggforce |
| gghalves | 0.1.4 | https://github.com/erocoar/gghalves |
| ggtree | 3.6.0 | https://bioconductor.org/packages/release/bioc/html/ggtree.html |
| htmltools | 0.5.7 | https://rstudio.github.io/htmltools/ |
| igraph | 1.6.0 | https://igraph.org/ |
| InterProScan | 5.64-96.0 | https://interproscan-docs.readthedocs.io/en/latest/ |
| iqtree | IQ-TREE multicore version 2.2.6 COVID-edition | http://www.iqtree.org/ |
| Kallisto | 0.48.0 | https://pachterlab.github.io/kallisto/ |
| limma | 3.58.1 | https://bioconductor.org/packages/release/bioc/html/limma.html |
| MAFFT | 7.520 | https://mafft.cbrc.jp/alignment/software/ |
| msaR | 0.6.0 | https://cran.r-project.org/web/packages/msaR/index.html |
| MultiQC | 1.16 | https://multiqc.info/ |
| Orthofinder | 2.5.5 | https://github.com/davidemms/OrthoFinder |
| pheatmap | 1.0.12 | https://cran.r-project.org/web/packages/pheatmap/index.html |
| plotly | 4.10.3 | https://plotly.com/r/ |
| Python | various versions were installed using conda for each software | https://www.python.org/ |
| qsmooth | 1.18.0 | https://github.com/stephaniehicks/qsmooth |
| R | 4.3.2 | https://cran.r-project.org/ |
| RColorBrewer | 1.1-3 | https://cran.r-project.org/web/packages/RColorBrewer/index.html |
| reshape | 0.8.9 | https://cran.r-project.org/web/packages/reshape/index.html |
| rmarkdown | 2.25 | https://rmarkdown.rstudio.com/ |
| RStudio | 2023.12.1 Build 402 | https://posit.co/download/rstudio-desktop/ |
| scales | 1.3.0 | https://cran.r-project.org/web/packages/scales/index.html |
| seqtk | 1.4-r122 | https://github.com/lh3/seqtk |
| Shiny | 1.8.0 | https://shiny.posit.co/ |
| shinydashboard | 0.7.2 | https://rstudio.github.io/shinydashboard/ |
| Sliding Window Inference for Network Generation (SWING) | - | https://github.com/bagherilab/SWING |
| Snakemake | 7.25.2 | https://snakemake.readthedocs.io/en/stable/index.html |
| TapScan | 2 | https://plantcode.cup.uni-freiburg.de/tapscan/ |
| The Gene Ontology Resource | - | https://geneontology.org/ |
| The Arabidopsis Information Resource | Araport11 | https://www.arabidopsis.org/ |
| Tidyverse | 2.0.0 | https://www.tidyverse.org/ |
| treeio | 1.22.0 | https://github.com/YuLab-SMU/treeio |
| Trimmomatic | 0.39 | http://www.usadellab.org/cms/index.php?page=trimmomatic |
| tximport | 1.30.0 | https://bioconductor.org/packages/release/bioc/html/tximport.html |
| Weighted Gene-Co-expression Network Analysis (WGCNA) | 1.72-5 |  https://web.archive.org/web/20221130104830/http://horvath.genetics.ucla.edu/html/CoexpressionNetwork/Rpackages/WGCNA/ |